### PR TITLE
feat: add target balances

### PR DESCRIPTION
**Pull request by mrice32**
This is a draft of how the target balance functionality might work. This is meant as an implementation proposal roughly following [this spec](https://docs.google.com/document/d/1kpu4r9FGU8CmAD6oUHVWX6V7DGkHww-hBdcdU7bQMLE/edit?usp=sharing).

The associated UMIP change is here: https://github.com/UMAprotocol/UMIPs/pull/556.

Note: the functionality is intended to be 100% backwards compatible, because a default threshold and target of 0 should result in the current behavior.

### CODE REVIEW

File: src/clients/ConfigStoreClient.ts [hunk 1]
```diff
@@ -85,6 +95,17 @@ export class AcrossConfigStoreClient {
     return config.transferThreshold;
   }
 
+  getSpokeTargetBalancesForBlock(
+    l1Token: string,
+    blockNumber: number = Number.MAX_SAFE_INTEGER
+  ): { [chainId: number]: SpokePoolTargetBalance } {
+    const config = (sortEventsDescending(this.cumulativeTokenTransferUpdates) as SpokeTargetBalanceUpdate[]).find(
+      (config) => config.blockNumber <= blockNumber && config.l1Token === l1Token
```
Comment (nicholaspai) [inline src/clients/ConfigStoreClient.ts:103]:
```suggestion
      (config) => config.l1Token === l1Token && config.blockNumber <= blockNumber
```

tiny nit but small optimization to filter on l1 token first. I doubt there will be 1000s of this event but just good practice to filter on rarer conditions first

File: src/clients/ConfigStoreClient.ts [hunk 2]
```diff
@@ -141,7 +163,22 @@ export class AcrossConfigStoreClient {
...
+          transferThreshold: toBN(transferThresholdForToken),
+          l1Token,
+        });
+
+        if (parsedValue?.spokeTargetBalances) {
+          // Note: cast is required because fromEntries always produces string keys, despite the function returning a
+          // numerical key.
+          const targetBalances = Object.fromEntries(
+            Object.entries(parsedValue.spokeTargetBalances).map(([chainId, targetBalance]) => {
+              return [chainId, { target: toBN(targetBalance.target), threshold: toBN(targetBalance.threshold) }];
```
Comment (nicholaspai) [inline src/clients/ConfigStoreClient.ts:177]: So targets will be in wei? I think this is consistent with the other parameter units like `tokenTransferthreshold`

File: src/dataworker/DataworkerUtils.ts [hunk 3]
```diff
@@ -151,9 +151,15 @@ export function _buildRelayerRefundRoot(
         tokenTransferThresholdOverrides[l1TokenCounterpart] ||
         clients.configStoreClient.getTokenTransferThresholdForBlock(l1TokenCounterpart, endBlockForMainnet);
 
+      const spokePoolTargetBalance = clients.configStoreClient.getSpokeTargetBalancesForBlock(
+        l1TokenCounterpart,
+        endBlockForMainnet
+      )[Number(repaymentChainId)] || { target: toBN(0), threshold: toBN(0) };
```
Comment (nicholaspai) [inline src/dataworker/DataworkerUtils.ts:157]: Should we do the defaulting (`|| { target: toBN(0), threshold: toBN(0) }`) in the ConfigStoreClient? Might be cleaner to abstract the defaulting logic away from the Dataworker

File: src/dataworker/PoolRebalanceUtils.ts [hunk 4]
```diff
@@ -285,6 +286,14 @@ export function constructPoolRebalanceLeaves(
             configStoreClient.getTokenTransferThresholdForBlock(l1Token, latestMainnetBlock)
         );
 
+        const spokeTargetBalances = l1TokensToIncludeInThisLeaf.map(
+          (l1Token) =>
+            configStoreClient.getSpokeTargetBalancesForBlock(l1Token, latestMainnetBlock)[Number(chainId)] || {
+              target: toBN(0),
```
Comment (nicholaspai) [inline src/dataworker/PoolRebalanceUtils.ts:292]: Same comment as above. I'm seeing a lot of this redundant defaulting code that could be removed

File: src/dataworker/PoolRebalanceUtils.ts [hunk 5]
```diff
@@ -316,16 +333,52 @@ export function constructPoolRebalanceLeaves(
...
+  // Transfer is always desired if hub owes spoke.
+  if (runningBalance.gte(0)) return runningBalance;
+
+  // Running balance is negative, but its absolute value is less than the spoke pool target balance threshold.
+  // In this case, we transfer nothing.
+  if (runningBalance.abs().lt(spokePoolTargetBalance.threshold)) return toBN(0);
+
+  // We are left with the case where the spoke pool is beyond the threshold.
+  // A transfer needs to be initiated to bring it down to the target.
+  const transferSize = runningBalance.abs().sub(spokePoolTargetBalance.target);
```
Comment (nicholaspai) [inline src/dataworker/PoolRebalanceUtils.ts:351]: This makes me think: should we enforce somewhere that the `spokePoolTargetBalance.target` is always positive?

File: src/dataworker/PoolRebalanceUtils.ts [hunk 6]
```diff
@@ -316,16 +333,52 @@ export function constructPoolRebalanceLeaves(
   return leaves;
 }
 
+// Note: this function computes the intended transfer amount before considering the transfer threshold.
+// A positive number indicates a transfer from hub to spoke.
+export function computeDesiredTransferAmount(
```
Comment (nicholaspai) [inline src/dataworker/PoolRebalanceUtils.ts:338]: Should we rename this function to `...ToSpoke` or `...toHub`? I'm a bit confused reading this in isolation to determine which direction the transfer amount should be interpreted as

**Review by nicholaspai — COMMENTED** [review]
This is really nice and overall very easy to understand, great work. Should we also open a draft UMIP PR reflecting these changes so they can be merged together?

File: src/clients/ConfigStoreClient.ts (hunk 1 above)
Comment (mrice32) [reply src/clients/ConfigStoreClient.ts:103]: Good point!

File: src/clients/ConfigStoreClient.ts (hunk 2 above)
Comment (mrice32) [reply src/clients/ConfigStoreClient.ts:177]: Yep, I think that makes the most sense and is consistent.

File: src/dataworker/DataworkerUtils.ts (hunk 3 above)
Comment (mrice32) [reply src/dataworker/DataworkerUtils.ts:157]: Hmmm, yeah, I think that could make sense. Let me try to rework it a little.

File: src/dataworker/PoolRebalanceUtils.ts (hunk 6 above)
Comment (mrice32) [reply src/dataworker/PoolRebalanceUtils.ts:338]: Good idea!

File: src/dataworker/DataworkerUtils.ts (hunk 3 above)
Comment (mrice32) [reply src/dataworker/DataworkerUtils.ts]: Done.

File: src/dataworker/PoolRebalanceUtils.ts (hunk 4 above)
Comment (mrice32) [reply src/dataworker/PoolRebalanceUtils.ts]: Done.

File: src/dataworker/PoolRebalanceUtils.ts (hunk 5 above)
Comment (mrice32) [reply src/dataworker/PoolRebalanceUtils.ts:347]: We have some choices:
- Ignore the specific target balance object with negative values.
- Ignore the entire updates that contain any negative values.
- Treat negative values as 0.

I've gone with the last option, but open to the other two as well -- they just might be a bit more complex. This has been added to the implementation. PTAL!

File: src/dataworker/PoolRebalanceUtils.ts (hunk 6 above)
Comment (mrice32) [reply src/dataworker/PoolRebalanceUtils.ts]: Done!

**Comment by mrice32** [comment]
> This is really nice and overall very easy to understand, great work. Should we also open a draft UMIP PR reflecting these changes so they can be merged together?

Yes, I will do this.

**Comment by mrice32** [comment]
@nicholaspai see the description for a link to the UMIP PR.

File: src/dataworker/PoolRebalanceUtils.ts (hunk 5 above)
Comment (nicholaspai) [reply src/dataworker/PoolRebalanceUtils.ts:347]: negative options as 0 makes sense!

**Review by nicholaspai — APPROVED** [review]
LGTM since this seems like it is backwards compatible and will not impact bots in prod without a config change on the ConfigStore contract

**Pull request merged**